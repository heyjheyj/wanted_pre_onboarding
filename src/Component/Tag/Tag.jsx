import React, { useState } from "react";
import TagComponent from "./ section/tagComponent";
import styles from "./Tag.module.css";

const Tag = ({ initTags }) => {
  const [tags, setTags] = useState(initTags);
  const [id, setId] = useState(3);

  const onSubmit = e => {
    if (e.key === "Enter" && e.target.value !== "") {
      setTags([...tags, { id: id, text: e.target.value }]);
      setId(id => id + 1);
      e.target.value = "";
    }
  };

  const deleteTag = tag => {
    setTags(tags.filter(item => item.id !== tag.id));
  };

  return (
    <div className={styles.container}>
      <h3 className={styles.title}>Tag</h3>
      <div className={styles.tagContainer}>
        <ul className={styles.tags} data-testid="test-taglist">
          {tags.map(tag =>
            <TagComponent key={tag.id} tag={tag} deleteTag={deleteTag} />
          )}
        </ul>
        <input
          className={styles.textInput}
          onKeyDown={onSubmit}
          data-testid="test-input"
          placeholder="Please enter to add tags"
        />
      </div>
    </div>
  );
};

export default Tag;
