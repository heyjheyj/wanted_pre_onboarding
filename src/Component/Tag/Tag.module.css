.container {
    width: 80%;
    height: 200px;
    border: 1px solid grey;
    margin-top: 10px;
    position: relative;
    display: flex;
    justify-content: center;
    align-items: center;
    border-radius: 10px;
}

.title {
    top: 5px;
    left: 5px;
    position: absolute;
}

.tagContainer {
    width: 80%;
    height: auto;
    border: 1px solid grey;
    border-radius: 10px;
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    align-items: flex-start;
    padding: 0 0 5px 0;
    max-height: 145px;
    overflow: scroll;
}

.tagContainer:focus-within {
    border: 1px solid #431cc6;
  }

.tags {
    width: auto;
    display: flex;
    flex-direction: row;
    align-items: center;
    flex-wrap: wrap;
}

.textInput {
    margin: 6px 0 0 6px;
    font-size: 14px;
    width: auto;
    border: none;
    border-radius: 10px;
    height: 32px;
}

input:focus {
    outline: none;
}
