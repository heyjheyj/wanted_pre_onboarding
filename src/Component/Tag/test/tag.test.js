import { render, screen } from "@testing-library/react";
import React from "react";
import TestRenderer from "react-test-renderer";
import Tag from "../Tag";
import userEvent from "@testing-library/user-event";

describe("Tag", () => {
  it("renders", () => {
    let initTags = [
      { id: 1, text: "CodeStates" },
      { id: 2, text: "jjang" }
    ];
    const component = TestRenderer.create(<Tag initTags={initTags} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe("Create a Tag", () => {
    let initTags = [
      { id: 1, text: "CodeStates" },
      { id: 2, text: "jjang" }
    ];
    let taglist;
    let input;
    let deletebutton1;
    let deletebutton2;

    beforeEach(() => {
      render(<Tag initTags={initTags} />);
      taglist = screen.getByTestId("test-taglist");
      input = screen.getByTestId("test-input");
      deletebutton1 = screen.getByTestId("test-deletebutton1");
      deletebutton2 = screen.getByTestId("test-deletebutton2");
    });

    it("creates a tag when user types something on input", () => {
      const newTag = "Good to see you";
      userEvent.type(input, newTag);
      userEvent.type(input, "{enter}");
      const createdTag = taglist.children[2];

      expect(taglist.children.length).toBe(3);
      expect(createdTag.childNodes[0].textContent).toBe("Good to see you");
    });

    it("deletes a tag when user clicks 'x' button in tag", () => {
      userEvent.click(deletebutton1);
      userEvent.click(deletebutton2);

      expect(taglist.children.length).toBe(0);
    });
  });
});
