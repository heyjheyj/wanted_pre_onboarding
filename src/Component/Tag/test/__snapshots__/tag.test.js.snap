// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag renders 1`] = `
<div
  className="container"
>
  <h3
    className="title"
  >
    Tag
  </h3>
  <div
    className="tagContainer"
  >
    <ul
      className="tags"
      data-testid="test-taglist"
    >
      <li
        className="tagcomponent"
        data-testid="test-tag1"
      >
        <span
          data-testid="test-text1"
        >
          CodeStates
        </span>
        <button
          data-testid="test-deletebutton1"
          onClick={[Function]}
        >
          x
        </button>
      </li>
      <li
        className="tagcomponent"
        data-testid="test-tag2"
      >
        <span
          data-testid="test-text2"
        >
          jjang
        </span>
        <button
          data-testid="test-deletebutton2"
          onClick={[Function]}
        >
          x
        </button>
      </li>
    </ul>
    <input
      className="textInput"
      data-testid="test-input"
      onKeyDown={[Function]}
      placeholder="Please enter to add tags"
    />
  </div>
</div>
`;
